flexget/plugins/filter/upgrade.py

```diff
@@ -0,0 +1,185 @@
+        grouped_entries[identifier.lower()].append(entry)
+
+    return grouped_entries
+
+
+class LazyUpgrade(object):
+
+    schema = {
+        'oneOf': [
+            {'type': 'boolean'},
+            {
+                'type': 'object',
+                'properties': {
+                    'identified_by': {'type': 'string'},
+                    'tracking': {'type': 'boolean'},
+                    'target': {'type': 'string', 'format': 'quality_requirements'},
+                    'on_lower': {'type': 'string', 'enum': ['accept', 'reject', 'fail', 'skip']}
+                },
+                'additionalProperties': False
+            }
+        ]
+    }
+
+    def prepare_config(self, config):
+        if not config or isinstance(config, bool):
+            config = {}
+        config.setdefault('identified_by', 'auto')
+        config.setdefault('tracking', True)
+        config.setdefault('on_lower', 'skip')
+        config.setdefault('target', None)
+        return config
+
+    def on_task_filter(self, task, config):
+        if not config:
+            return
+
+        config = self.prepare_config(config)
+
+        grouped_entries = group_entries(task.entries, config['identified_by'])
+
+        with Session() as session:
+            # Prefetch Data
+            existing_ids = session.query(EntryUpgrade).filter(EntryUpgrade.id.in_(grouped_entries.keys())).all()
+            existing_ids = {e.id: e for e in existing_ids}
+
+            for identifier, entries in grouped_entries.items():
+                if not entries:
+                    continue
+
+                existing = existing_ids.get(identifier)
+                if not existing:
+                    # No existing, skip
+                    continue
+
+                target_downloaded = False
+
+                if config['target']:
+                    target_quality = qualities.Quality(config['target'])
+                    target_requirement = qualities.Requirements(config['target'])
+                    # Filter out entries within target
+                    entries = list(filter(lambda e: target_requirement.allows(e['quality']), entries))
+
+                    # Have we already downloaded target
+                    if existing.quality > target_quality or target_requirement.allows(existing.quality):
+                        continue
+
+                # We may have no entries within target
+                if len(entries) == 0:
+                    continue
+
+                # Sort entities in order of quality
+                entries.sort(key=lambda e: e['quality'], reverse=True)
+
+                # First entry will be the best quality
+                if entries[0]['quality'] > existing.quality:
+                    # Accept if we have a better quality
+                    entries[0].accept('upgraded quality')
+
+                # Action lower qualities
+                if config['on_lower'] != 'skip':
+                    for entry in entries:
```

Comment:
shouldnt you pop the first entry to not iterate over it twice?

Reply:
not sure it really matters?

Reply:
well i think itll try to accept it twice, not sure if that's an issue or not